### PR TITLE
Fix race in local storage

### PR DESCRIPTION
**Pull request by zeripath**
LocalStorage should only put completed files in position

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
There are a couple of other races in here - which I don't really know how to fix.

The issues are:

* A broken upload will temporarily appear in the store before being taken out but could be downloaded in the intervening period - and you could actually delete a good upload whilst deleting a bad one.
* A broken object in the store being noticed at download time isn't very easy to fix either.

**Comment by zeripath** [comment]
This PR however does get rid of the most egregious problem.

**Review by 6543 — APPROVED** [review]

File: modules/storage/local.go [hunk 1]
```diff
@@ -64,17 +72,37 @@ func (l *LocalStorage) Save(path string, r io.Reader) (int64, error) {
...
 		return 0, err
 	}
+	tmp, err := ioutil.TempFile(l.tmpdir, "upload-*")
+	if err != nil {
+		return 0, err
+	}
+	tmpRemoved := false
+	defer func() {
+		if !tmpRemoved {
+			util.Remove(tmp.Name())
```
Comment (zeripath) [inline modules/storage/local.go]:
```suggestion
			_ = util.Remove(tmp.Name())
```

File: modules/storage/local.go [hunk 2]
```diff
@@ -46,9 +49,14 @@ func NewLocalStorage(ctx context.Context, cfg interface{}) (ObjectStorage, error
 		return nil, err
 	}
 
+	if config.TemporaryPath == "" {
```
Comment (lunny) [inline modules/storage/local.go:52]: Put them to a system temp directory is better.

Comment (zeripath) [reply modules/storage/local.go:52]: These files may be large and the system temporary directory is not guaranteed to be able to cope. 

The system tempdir may be on a different volume in Windows or may be on a different filesystem.

The only place we know for sure that LFS files will fit and should be renamed atomically is if they're in the LFS repository path.

The temporary directory is configurable - although I've not explicitly documented it here: set TEMPORARY_PATH for the storage

**Comment by zeripath** [comment]
Fundamentally the other races exist because there is a verification step that the storage system is not aware of and isn't doing.

But thinking on I suppose we could "buffer" the data being passed to the storage system so that by the time eof is called we know that the data we're passing has actually got the right SHA256 and is the right size. 

That is: instead of using a teereader we wrap the reader and write to the hash and keep account of the size during the Read(...)

Then when EOF is returned from the body reader we check the hash and the size - determine if it's correct before returning io.EOF otherwise we return some other error. 

The other error will then percolate up to the storage reader and its io.copy killing them and aborting the storage. Boom. 

Verification in stream and no more races.

**Comment by zeripath** [comment]
Now it's likely that the body reader implements io.WriterTo which may mean we need to implement io.Writer in addition to WriterTo to do the same trick just slightly differently.

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
Backport -> #14901

**Comment by zeripath** [comment]
Sorry was in bed so couldn't do the backport!

**Pull request merged**